File: examples/csharp/Logos.PaymentService.Application/UseCases/GetPaymentUseCase.cs
using Logos.PaymentService.Application.Contracts;
using Logos.PaymentService.Domain.Abstractions;

namespace Logos.PaymentService.Application.UseCases;

public class GetPaymentUseCase(IPaymentRepository paymentRepository)
{
    public GetPaymentResponse? Execute(GetPaymentRequest request)
    {
        var payment = paymentRepository.GetById(request.PaymentId);

        if (payment == null)
            return null;

        return new GetPaymentResponse(
            payment.Id,
            payment.Amount,
            payment.MerchantId,
            payment.Status,
            payment.CreatedAt,
            payment.DeclineReason);
    }
}

File: examples/cpp/logos_payment_service_application/use_cases/get_payment_use_case.cpp
#include "logos_payment_service_application/use_cases/get_payment_use_case.h"

namespace logos::payment_service::application::use_cases {

GetPaymentUseCase::GetPaymentUseCase(domain::abstractions::IPaymentRepository* repository)
    : repository_(repository) {
}

std::optional<contracts::GetPaymentResponse> 
GetPaymentUseCase::Execute(const std::string& payment_id) {
    return repository_->GetById(payment_id);
}

} // namespace logos::payment_service::application::use_cases
